### ISSUE TITLE
boost python setup 

### ISSUE THREAD

**Issue by pavlis** [wangyinz/mspass#19]
The purpose of posting this issue is to document some problems I've encountered in building boost python that we are using to construct C++ wrapper for python.   Some of these issues might disappear with time, but not necessarily.

The first issue is it seems the documentation for boost python is out of date by a fair number of years.   That is a bit disturbing as it may indicate a lack of support for this package that could become an issue down the road.  The evidence for this is the documentation only discusses python versions up to about 2.4 or 2.5s.   Notes and [this article](https://www.boost.org/doc/libs/1_71_0/libs/python/doc/html/article.html) suggest most of this was developed more than a decade ago.   We may want to make sure this package has a significant following before we go too far down this path.
Some issues I encountered:

1. The documentation talks about "bjam".  I am 99% sure bjam was, at some time in the past, change to "b2" that presumably has different features.
2.  Evidence of this is that the quickstart demo they give for testing boost python uses arguments to b2 (they call it bjam) that have changed.   
3.  I think there are errors in the default configuration for b2 to build this example as I cannot make this error go away: /usr/bin/ld: cannot find -lboost_python.   I am about 99% sure it should be trying to link to -lboost_python27.   If I could figure out how to change the python default to a 3 version I would, but haven't cracked this yet.   

Not ready to give up on this yet, but we might want to do some poking around to make sure we aren't mounting a donkey instead of getting into a Tesla.

**Comment by pavlis** [comment]
I think we need to look seriously at [cython](http://docs.cython.org/en/latest/index.html).  Looks to me that this can solve our problems with a much cleaner approach than boost python.   
Cython is a superset of python that can be compiled and link directly with C programs.   This could make a huge difference in speed for some applications as well.   Ian, why don't you look at this and let's plan to talk about this in a phone call soon.

**Comment by wangyinz** [comment]
I initially was going to say Cython would not fit our purpose here. Then, I read their documentation, and found that it is more than just a Python-to-C interface now, and it got very detailed documentation on how to use it with C++. I think that looks pretty promising. 

The other package that I have come across before playing with Boost.Python is the [pybind11](https://github.com/pybind/pybind11). It is similar to Boost.Python, but designed to be more user-friendly. I have not really tried it, but the tool that I use to generate the mangled binding actually supports pybind11, too. That's definitely another alternative we could try. 

I think one way to find out whether a package is gonna survive is to look at all the stats on their GitHub page (i.e. latest commits, stars, forks). It seems to me that the Boost.Python repo is at a diminishing stage, and there is not much significant updates recently. Both Cython and pybind11 are very active in all aspects. I guess that does imply something about the longevity of the projects. I originally chose Boost.Python because we are already using Boost in our project, and the less dependency usually translates to more stability, and Boost is definitely not going away in the foreseeable future, but I guess Boost.Python is not necessarily the indispensable component of Boost.

**Comment by pavlis** [comment]
I think I'm going to pursue cython a bit.   There is at lest one case cython doesn't seem to handle that may make a wrapper approach like boost python on pybind11 better.  That is, creating a subclass of a pure C++ object in cython appears from the documentation to be problematic.   Might be possible, but the combination of little mention of the issue other than a warning of limitations suggests it would be an issue.   Comes up immediately as a Seismogram and TimeSeries implementation in cython would be really slick.   It would allow constructors from obspy objects or with a database handle.   Without that we'd need a more awkward interface.   On the other hand, with conventional wrappers that would be possible but very hard to do.   

I'm going to play with cython a bit today and see if I can determine if I can create a TimeSeries class that inherits my (revised) CoreTimeSeries and has an obspy constructor.

**Comment by pavlis** [comment]
After writing that last comment I followed your link to pybind11 and changed my mind.    It is basically a new and improved boost python.   It solves a problem that was concerning me even without trying it.  That is, with boost python it looked very awkward to try to crack a standard python container like a dict or list in C.   pybind11 has "thin wrappers" that make that trivial at the application level.   I'm going to play with pybind11 a bit.   Incidentally the 11 in the name comes about because the the package is simplified from boost python by taking advantage of new features in C11

**Comment by pavlis** [comment]
For the record, pybind11 was pretty easy to install on ubuntu, BUT there were a bunch of undocumented dependencies.  I'm not sure how important these dependencies are, but they were:
1.  When cmake ran it warned that eigen was missing.  It completed anyway suggesting this is not critical.   I installed it anyway, reran cmake, and the error went away.
2.   Not surprisingly the package requires a lot of python packages.  I had to first install pytest.  The this long string was needed as my python version was bare bones on this new ubuntu box:
 python -m pip install --user numpy scipy matplotlib ipython jupyter pandas sympy nose
Note that ubuntu apt-get doesn't work for installing as it throws an error about ipython being required but there is no package for it.    The documentation on the above line mentions that they suggest --user because installing some of these as root is dangerous.  

This is very relevant as if we settle on pybind11 our docker image will need these things.

**Comment by pavlis** [comment]
I've been struggling to get pybind11 to work and I'm about to throw up my hands and try out cython as an alternative.   Classic thing were you look at something on the surface at it looks pretty clean, but when you get into the weeds it become bewildering.   I found it simple to write wrappers for simple classes that didn't do anything fancy.  However, after writing in C++ as long as I have  I've learned to do a few "fancy" things that turn out to be poorly documented and very confusing (at least to me).   Some top level problems are:

1. The weird and confusing syntax required for the wrapper code.  
2.  Lots of hidden defaults that left me scratching my head and doing a lot of trial and error hacking.   
3.  Because the whole suite is built from C++ templates it can be really hard to figure out why something throws an error.   Simple typos can create pages of compile message. 

So, wrapping a simple class like my SlownessVector was quite trivial and required minimal knowledge.  Where I quickly got into the weeks was trying wrap Metadata, which is a core thing we need in this project.  Things it has that require strange incantations are:

1. Function overloading.   They have a clean approach, but it takes a whole section to describe it and the cleanest solution requires C++14.
2.  Inheritance from an abstract base class.  Something the documentation never explicitly even addresses, but turns out to be an end member of the thorny problem in this package (boost would be too I think) of wrapping virtual functions of any kind.   I finally implemented a version of  a "trampoline" class  after hours of struggling.   The syntax is odd and there are insufficient examples around on the web that could help me figure this out.  I did finally get it to compile, but at the moment I can't make the python part work (more on that in a minute). 
3.   Handling "const" things, I think, are giving me trouble.  const args are easy as long as you get them right, but I got some mysterious messages from compilation when I missed a const qualifier.   I eventually found a paragraph in the documentation that notes that python as no notion of const to they admit there have a problem on things related to that issue.  

At the moment I'm stuck on something I find bewildering.   I can create a valid library I can import correctly.   I can run something like this:
import mspasspy
m=mspasspy.Metadata

But then if I try something like this:
m.put('foo',2.0)
I get a "TypeError" message even when it echos that it tried the overloaded put methods with signature 'str' and 'double'   That is where  I'm hung up right now.  Web searches have not helped as TypeError seems come from a variety of very different issues.   

I may eventually figure this out and still use pybind11, but I think it will be useful to experiment with cython before I get too far down this road.  It si very clear inheritance issues are ugly with pybind11 and since all the main data objects in my library use not just regular inheritance, but multiple inheritance I worry about the long term maintainability of going the pybind11 route. So, am punting to cython for a second option next.

**Comment by pavlis** [comment]
After some playing around with cython I am 100% sure this should be our primary tool for integrating C and python.   The reason is less the problem noted in the previous post with pybind11 difficulties, but more with the functionality cython will provide for the framework.   The issue is that I forsee we could recommend in the documentation this model for developing a new algorithm:

1. Develop and test a python module that could plug into spark.
2.  Test the python code running through sparc if parallelization is needed
3.  cythonize the code if necessary for speed.   
4.  If profiling showed bottlenecks in the cython implementation implement sections of the algorithm in C or C++ to further improve performance.

This represents a specific example of an old axiom I learned decades ago:  make it work before you make it fast.

Although  it has it's own idiosyncrasies Cython looks a bit more straightforward for most applications than pybind11.  Here are two key issues I need to resolve before I can completely recommend this route:

1. I have not found anything in the documentation yet about dealing with inheritance and polymorphism.   I am going on the assumption that a wrapper for a class like TimeSeries needs to define all methods it should expose to python.   This means different classes that share another by inheritance (e.g. Metadata) will require cut and paste of common method definitions.  If so, that is fine, but I need to test that theory.   I find it odd that search for" "polymorphism in cython" or similar do not yield any information in google.   
2.  How to automate compilation with cython using cmake is not yet clear.  I found an excellent example on github thanks to google so don't think this is a major barrier, but won't be trivial either.  Initially I'm just using the documents manual build system using distutil and a file called setup.py that the documentation indicates is the required method.   I just haven't quite figured out how the cmake configuration works with that approach.

**Comment by wangyinz** [comment]
For the pybind11 example you gave, I think I know what the TypeError means. I think you should do the following instead:
```
import mspasspy
m=mspasspy.Metadata()
m.put('foo',2.0)
```
With the parenthesis after Metadata, the `m` is constructed as type `Metadata`. Without that, it is basically assigning an alias. The issue could be more than this, but hope that helps.

I think the development workflow you proposed for the user makes sense. Maybe Cython is the cleaner solution for us. Anyway, you knows it much better, so I will leave you to decide. I can definitely help you fix the issues with CMake.

**Comment by pavlis** [comment]
I'm going to end this chain here and move back to the issue titled "how to smoothly get python and C code to interact".   boost python is out.   pybind11 is maybe, but I had a duhh moment about cython discussed in that issues section.

**Comment by pavlis** [comment]
The above was not the end of this chain.   I had pushed away the pybind11 branch and was playing with the cython alternative when I wrote that comment.   Wanted to point out here that your suggestion, Ian, to use m=mspasspy.Metadata() did the trick and the TypeError problem disappeared.   What a strange issue of a subtle difference in syntax.  It is, however, typical.

I am still moving the continuation of this conversation to "how to smoothly get python and C code to interact", but thought this was the right place to record this important trick.
